core/services/workflow_service/controllers/workflow_controller.py

```diff
@@ -0,0 +1,340 @@
+from services.workflow_service.controllers import compute_block_controller
+from services.workflow_service.schemas.workflow import (
+    WorfklowValidationError,
+    BlockStatus
+)
+from airflow_client.client import ApiClient, Configuration, ApiException
+from airflow_client.client.api.dag_run_api import DAGRunApi
+from airflow_client.client.model.list_dag_runs_form import ListDagRunsForm
+from airflow_client.client.model.dag_run import DAGRun
+from airflow_client.client.api.dag_api import DAGApi
+from airflow_client.client.api.task_instance_api import TaskInstanceApi
+from airflow_client.client.model.dag import DAG
+
+DAG_DIRECTORY = ENV.AIRFLOW_DAG_DIR
+airflow_config = Configuration(
+    host=ENV.AIRFLOW_HOST,
+    username=ENV.AIRFLOW_USER,
+    password=ENV.AIRFLOW_PASS
+)
+
+
+def _project_id_to_dag_id(pi: UUID | str) -> str:
+    return f"dag_{str(pi).replace("-", "_")}"
+
+
+def _dag_id_to_project_id(di: str) -> str:
+    return di.replace("dag_", "").replace("_", "-")
+
+
+def _task_id_to_cb_id(ti: str) -> str:
+    return ti.replace("task_", "").replace("_", "-")
+
+
+def _cb_id_to_task_id(ci: UUID | str) -> str:
+    return f"task_{str(ci).replace('-', '_')}"
+
+
+def parse_configs(configs):
+    return {k: json.dumps(v) if isinstance(v, list) else str(v)
+            for k, v in configs.items()}
+
+
+def create_graph(project):
+    graph = nx.DiGraph()
+
+    for block in project.blocks:
+        entrypoint = block.selected_entrypoint
+        configs = [io.config for io in entrypoint.input_outputs]
+        merged_configs = {
+            **parse_configs(entrypoint.envs),
+            **{k: v for d in configs for k, v in parse_configs(d).items()}
+        }
+
+        graph.add_node(block.uuid, **{
+            "uuid": block.uuid,
+            "name": block.name,
+            "image": block.docker_image,
+            "entry_name": block.selected_entrypoint.name,
+            "environment": merged_configs,
+        })
+
+    # Add edges (dependencies)
+    for block in project.blocks:
+        for upstream in block.upstream_blocks:
+            graph.add_edge(upstream.uuid, block.uuid)
+
+    # Ensure the graph is a valid DAG
+    if not nx.is_directed_acyclic_graph(graph):
+        raise HTTPException(
+            status_code=400, detail="The project is not acyclic.")
+
+    if not nx.is_connected(graph.to_undirected()):
+        raise HTTPException(
+            status_code=400, detail="Not all compute blocks are connected."
+        )
+
+    return graph
+
+
+def init_templates():
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    templates_dir = os.path.join(base_dir, "..", "templates")
+
+    env = Environment(loader=FileSystemLoader(templates_dir))
+    return {
+        "dag": env.get_template("dag_base.py.j2"),
+        "algorithm": env.get_template("algorithm_docker.py.j2"),
+        "dependency": env.get_template("dependency.py.j2"),
+    }
+
+
+def gen_dag_code(graph, templates, dag_id, project_uuid):
```

Comment:
```suggestion
def generate_dag_code(graph, templates, dag_id, project_uuid):
```

Be more explicit with the function names. It does not hurt anyone 😉